### PR TITLE
Project Mongo API, Anna Robertsson

### PR DESCRIPTION
**Pull request by AnnaRobertsson**
https://mongoapi-fkkq.onrender.com

### CODE REVIEW

File: routes/songRoutes.js [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+  const popularity = parseInt(req.query.popularity, 10); // Parse the 'popularity' query parameter as an integer
+
+  // Find the album by title in the database
+  MetallicaSongModel.find({ album: { $regex: requestedAlbum, $options: "i" } })
+    .then((foundAlbum) => {
+      // If no album found, send a 404 error response
+      if (!foundAlbum || foundAlbum.length === 0) {
+        return res
+          .status(404)
+          .json({ message: "Album not found, please try another album title" });
```
Comment (emmydieden) [inline routes/songRoutes.js:94]: This error handling...

File: routes/songRoutes.js [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        // Filter songs within the album by popularity
+        const songsWithDesiredPopularity = foundAlbum.filter(
+          (song) => song.popularity >= popularity
+        );
+
+        // If no songs match the popularity criterion, send an error response
+        if (songsWithDesiredPopularity.length === 0) {
+          return res.status(404).json({
+            message:
+              "No songs found with this popularity or more in this album",
```
Comment (emmydieden) [inline routes/songRoutes.js:108]: ...and this error handling are well set up and guides the user in a clear way on what has gone wrong when searching for an album and songs with a certain popularity within it.

File: routes/songRoutes.js [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    })
+    .catch((error) => {
+      // If an error occurs during the database query, send an error response
+      res
+        .status(500)
+        .json({ message: "Internal server error", error: error.message });
+    });
+});
+
+// ------- Dummy endpoint -------
```
Comment (emmydieden) [inline routes/songRoutes.js:126]: Is this commented out because the route isn't working?

**Review by emmydieden — COMMENTED** [review]
You have a great looking API Anna! As always you come up with fun data:) I think you did really well with your error handling for all the endpoints. They make it clear when potential error occur and make debugging simpler. 

I appreciate the guidance of endpoints you gave in the readme-file. Although, it would have been easier if you would have written the chunk of text in bullet points instead! Just a small detail:) And also - the example link you gave in the readme to try out filtering album and popularity didn't work. The album name is missing spaces between the words. 

Great job and thanks for letting me review your code this week!

File: routes/songRoutes.js [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    });
+});
+
+// ------- One Metallica song based on title -------
+// Show a specific Metallica song based on the song title (case-insensitive and partial match)
+router.get("/song-name/:title", (req, res) => {
+  const songTitle = req.params.title.toLowerCase(); // Convert to lowercase for case-insensitive search
+
+  // Find a song by its title using the Mongoose model, allowing partial matches
+  MetallicaSongModel.find({ name: { $regex: songTitle, $options: "i" } })
```
[the remark this answers is not in the document]
Comment (AnnaRobertsson) [reply routes/songRoutes.js:61]: The "i" makes it case-insensitive from my understanding :)

File: routes/songRoutes.js (hunk 3 above)
Comment (AnnaRobertsson) [reply routes/songRoutes.js:126]: It said that the previous weeks stretch goals were applicable here too (and using dummy endpoints was one of them). I didn't get this to fully work so I made it a dummy endpoint to showcase what you could have continued working on sort of..

**Pull request closed without merging**